packages/node/src/lib/log.ts

```diff
@@ -0,0 +1,145 @@
+import type { LogOptions } from './construct-payload';
+import type { GroupingObject, OutgoingLogBody } from './metrics-log';
+import type { Request, Response } from 'express';
+import config from '../config';
+import clamp from 'lodash/clamp';
+import * as url from 'url';
+import { v4 as uuidv4 } from 'uuid';
+import { constructPayload } from './construct-payload';
+import { metricsAPICall } from './metrics-log';
+
+let queue: OutgoingLogBody[] = [];
+function doSend(readmeApiKey, options) {
+  // Copy the queue so we can send all the requests in one batch
+  const json = queue.slice();
+  // Clear out the queue so we don't resend any data in the future
+  queue = [];
+
+  // Make the log call
+  metricsAPICall(readmeApiKey, json).catch(e => {
+    // Silently discard errors and timeouts.
+    if (options.development) throw e;
+  });
+}
+// Make sure we flush the queue if the process is exited
+process.on('exit', doSend);
+
+interface ExtendedResponse extends Response {
+  _body?: string;
+}
+
+// We're doing this to buffer up the response body
+// so we can send it off to the metrics server
+// It's unfortunate that this isn't accessible
+// natively. This may take up lots of memory on
+// big responses, we can make it configurable in future
+function patchResponse(res) {
+  const { write, end } = res;
+
+  res._body = '';
+
+  res.write = (chunk, encoding, cb) => {
+    res._body += chunk;
+    write.call(res, chunk, encoding, cb);
+  };
+
+  res.end = (chunk, encoding, cb) => {
+    // Chunk is optional in res.end
+    // http://nodejs.org/dist/latest/docs/api/http.html#http_response_end_data_encoding_callback
+    if (chunk) res._body += chunk;
+    end.call(res, chunk, encoding, cb);
+  };
+}
+
+export interface Options extends LogOptions {
+  bufferLength?: number;
+  baseLogUrl?: string;
+}
+
+/**
+ * This middleware will set up Express to automatically log all API requests to ReadMe Metrics.
+ *
+ * @param apiKey The API key for your ReadMe project. This ensures your requests end up in your dashboard. You can read more about the API key in [our docs](https://docs.readme.com/reference/authentication).
+ * @param group A function that helps translate incoming request data to our metrics grouping data. You can read more under [Grouping Function](#grouping-function).
+ * @param options Additional options. See the documentation for more details.
+ * @returns Your Express middleware
+ */
+export function log(
+  readmeApiKey: string,
+  req: Request,
+  res: ExtendedResponse,
+  group: GroupingObject,
+  options: Options = {}
+) {
+  if (!readmeApiKey) throw new Error('You must provide your ReadMe API key');
+  if (!group) throw new Error('You must provide a group');
+
+  // Ensures the buffer length is between 1 and 30
+  const bufferLength = clamp(options.bufferLength || config.bufferLength, 1, 30);
+
+  const baseLogUrl = options.baseLogUrl || undefined;
```

Comment:
nit: I generally prefer using `null` instead of `undefined` when explicitly setting a value, something like `const { baseLogUrl = null } = options;`

Reply:
Yeah fair point, tbh I'm not sure what to do about this option overall. Idk whether it makes sense for us to be doing this for them or encouraging it via code sample instead: https://github.com/readmeio/metrics-sdks/issues/562